httpbp/client_middlewares.go

```diff
@@ -0,0 +1,222 @@
+	return f(req)
+}
+
+// NewClient returns a standard HTTP client wrapped with the default
+// middleware plus any additional client middleware passed into this
+// function.
+func NewClient(config ClientConfig, middleware ...ClientMiddleware) *http.Client {
+	transport := &http.Transport{
+		MaxConnsPerHost: config.MaxConnections,
+	}
+
+	defaults := []ClientMiddleware{
+		ClientErrorWrapper(),
+		MonitorClient(config.Slug),
+		Retries(
+			retry.Attempts(1),
+			retrybp.Filters(
+				retrybp.NetworkErrorFilter,
+				retrybp.RetryableErrorFilter,
+			),
+		),
+	}
+	if config.CircuitBreaker != nil {
+		defaults = append(defaults, CircuitBreaker(*config.CircuitBreaker))
+	}
+
+	middleware = append(middleware, defaults...)
+
+	return &http.Client{
+		Transport: WrapTransport(transport, middleware...),
+	}
+}
+
+// WrapTransport takes a list of client middleware and wraps them around the
+// given transport. This is useful for using client middleware outside of this
+// package.
+func WrapTransport(transport http.RoundTripper, middleware ...ClientMiddleware) http.RoundTripper {
+	if transport == nil {
+		transport = http.DefaultTransport
+	}
+	// add middleware in reverse so the first in the list is the outermost
+	for i := len(middleware) - 1; i >= 0; i-- {
+		transport = middleware[i](transport)
+	}
+	return transport
+}
+
+// ClientErrorWrapper applies ClientErrorFromResponse to the returned response
+// ensuring a HTTP status response outside the range [200, 400) is wrapped in
+// an error relieving users from the need to check the status response.
+//
+// Adding this middleware means that both, non-nil response and non-nil error,
+// can be returned. If configured the caller needs to call
+// httpbp.DrainAndClose(resp.Body) to ensure the underlying TCP connection can
+// be re-used.
+func ClientErrorWrapper() ClientMiddleware {
+	const limit = 1024
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripperFunc(func(req *http.Request) (resp *http.Response, err error) {
+			resp, err = next.RoundTrip(req)
+			if err != nil {
+				return resp, err
+			}
+			if err == nil {
+				err = ClientErrorFromResponse(resp)
+				if err != nil {
+					defer DrainAndClose(resp.Body)
+					var ce *ClientError
+					if !errors.As(err, &ce) {
+						return resp, err
+					}
+					body, e := io.ReadAll(io.LimitReader(resp.Body, limit))
+					if e != nil {
+						return resp, e
+					}
+					ce.AdditionalInfo = string(body)
+					return resp, ce
+				}
+			}
+			return resp, err
+		})
+	}
+}
+
+// CircuitBreaker is a middleware that prevents sending requests that are likely
+// to fail based on a configurable failure ratio based on total failures and
+// requests.
+func CircuitBreaker(config breakerbp.Config) ClientMiddleware {
+	pool := &sync.Pool{
+		New: func() interface{} {
+			return breakerbp.NewFailureRatioBreaker(config)
+		},
+	}
+	breakers := &sync.Map{}
```

Comment:
style nit: this doesn't need to be a pointer either, and as a result prefer `var breakers sync.Map` (https://github.com/uber-go/guide/blob/master/style.md#use-var-for-zero-value-structs)